### PR TITLE
Tessera Orion-mode privacy ATs

### PR DESCRIPTION
**Pull request by mark-terry**
## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java [hunk 1]
```diff
@@ -246,7 +246,7 @@ public void startNode(final BesuNode node) {
     }
 
     params.add("--Xsecp256k1-native-enabled=" + node.isSecp256k1Native());
-    params.add("--Xaltbn128-native-enabled=" + node.isAltbn128Native());
+    params.add("--Xaltbn128-native-enabled=false" + node.isAltbn128Native());
```
Comment (macfarla) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java:249]: what are the implications of this change?

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/PrivacyCluster.java [hunk 2]
```diff
@@ -145,7 +148,24 @@ private void startNode(final PrivacyNode node, final boolean isBootNode) {
         .ifPresent(node.getConfiguration()::setGenesisConfig);
 
     if (!isBootNode) {
-      node.addOtherEnclaveNode(bootNode.getOrion().nodeUrl());
+      if (bootNode.getEnvclave().getEnclaveType() == EnclaveType.TESSERA) {
```
Comment (macfarla) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/PrivacyCluster.java:151]: typo here s/envclave/enclave

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/DeployPrivateSmartContractAcceptanceTest.java [hunk 3]
```diff
@@ -31,14 +38,21 @@
...
-            "miner-node", privacyAccountResolver.resolve(0));
+            "miner-node", privacyAccountResolver.resolve(0), enclaveType, Optional.empty());
     privacyCluster.start(minerNode);
   }
 
   @Test
   public void deployingMustGiveValidReceiptAndCode() throws Exception {
     final String contractAddress = "0x89ce396d0f9f937ddfa71113e29b2081c4869555";
 
+    //    boolean wahr = true;
```
Comment (macfarla) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/DeployPrivateSmartContractAcceptanceTest.java:49]: cleanup :)

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/PrivacyGroupAcceptanceTest.java [hunk 4]
```diff
@@ -122,11 +145,12 @@ public void nodeCanCreatePrivacyGroupWithoutOptionalParams() {
             PrivacyGroup.Type.PANTHEON,
             "",
             "",
-            Base64String.wrapList(alice.getEnclaveKey(), bob.getEnclaveKey()));
+            Base64String.wrapList(alice.getEnclaveKey()));
+    //    Base64String.wrapList(alice.getEnclaveKey(), bob.getEnclaveKey()));
```
Comment (macfarla) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/PrivacyGroupAcceptanceTest.java:149]: cleanup

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/PrivacyGroupAcceptanceTest.java [hunk 5]
```diff
@@ -122,11 +145,12 @@ public void nodeCanCreatePrivacyGroupWithoutOptionalParams() {
...
 
     alice.verify(privateTransactionVerifier.validPrivacyGroupCreated(expected));
 
-    bob.verify(privateTransactionVerifier.validPrivacyGroupCreated(expected));
+    //    bob.verify(privateTransactionVerifier.validPrivacyGroupCreated(expected));
```
Comment (macfarla) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/PrivacyGroupAcceptanceTest.java:153]: cleanup

**Review by macfarla — COMMENTED** [review]
some comments

File: crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java [hunk 6]
```diff
@@ -66,7 +66,7 @@
 
   private static final Logger LOG = LogManager.getLogger();
 
-  private boolean useNative = true;
+  private boolean useNative = false;
```
Comment (macfarla) [inline crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java:69]: this doesn't seem right

File: enclave/src/integration-test/java/org/hyperledger/besu/enclave/EnclaveTest.java [hunk 7]
```diff
@@ -59,8 +59,9 @@ public void setUp() throws Exception {
 
     testHarness =
         OrionTestHarnessFactory.create(
+            "enclave",
             folder.newFolder().toPath(),
-            new OrionKeyConfiguration("orion_key_0.pub", "orion_key_0.key"));
+            new EnclaveKeyConfiguration("orion_key_0.pub", "orion_key_0.key"));
```
Comment (macfarla) [inline enclave/src/integration-test/java/org/hyperledger/besu/enclave/EnclaveTest.java:64]: should we rename the keys?

File: enclave/src/main/java/org/hyperledger/besu/enclave/EnclaveClientException.java [hunk 8]
```diff
@@ -14,7 +14,7 @@
  */
 package org.hyperledger.besu.enclave;
 
-public class EnclaveClientException extends IllegalArgumentException {
+public class EnclaveClientException extends RuntimeException {
```
Comment (macfarla) [inline enclave/src/main/java/org/hyperledger/besu/enclave/EnclaveClientException.java:17]: why this change?

File: enclave/src/main/java/org/hyperledger/besu/enclave/VertxRequestTransmitter.java [hunk 9]
```diff
@@ -73,6 +77,7 @@ public VertxRequestTransmitter(final HttpClient httpClient) {
               .exceptionHandler(result::completeExceptionally)
               .setChunked(false);
       contentType.ifPresent(ct -> request.putHeader(HttpHeaders.CONTENT_TYPE, ct));
+      LOG.info("Sending this request: " + request.absoluteURI() + request.method() + content.get());
```
Comment (macfarla) [inline enclave/src/main/java/org/hyperledger/besu/enclave/VertxRequestTransmitter.java:80]: TRACE or DEBUG?

**Review by macfarla — COMMENTED** [review]
some more comments

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java (hunk 1 above)
Comment (mark-terry) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java:249]: Not sure, but that looks broken. @pinges ?

Comment (macfarla) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java:249]: I wonder if this is a merge fail - there was a similar change here crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java [hunk 10]
```diff
@@ -176,6 +176,12 @@
   ENCLAVE_PAYLOAD_NOT_FOUND(-50200, "EnclavePayloadNotFound"),
   CREATE_GROUP_INCLUDE_SELF(-50200, "CreatePrivacyGroupShouldIncludeSelf"),
 
+  // Tessera error codes
+  TESSERA_NODE_MISSING_PEER_URL(-50200, "Recipient not found for key:"),
+  TESSERA_CREATE_GROUP_INCLUDE_SELF(
+      -50200, "The list of members in a privacy group should include self"),
+  // TESS(-50200, "Public key not found when searching for private key");
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java:183]: commented out code

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/EnclaveErrorAcceptanceTest.java [hunk 11]
```diff
@@ -174,7 +179,17 @@ public void transactionFailsIfPartyIsOffline() {
...
+    assertThat(throwable.getMessage())
+        .has(matchOrionOrTesseraMessage(orionMessage, tesseraMessage));
+  }
 
-    assertThat(throwable).hasMessageContaining(JsonRpcError.CREATE_GROUP_INCLUDE_SELF.getMessage());
+  private Condition<String> matchOrionOrTesseraMessage(
+      final String orionMessage, final String tesseraMessage) {
+    return new Condition<>(
+        message -> message.contains(orionMessage) || message.contains(tesseraMessage),
+        "Message did not match either Orion or Tessera expected output");
```
Comment (macfarla) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/EnclaveErrorAcceptanceTest.java:193]: Would be useful to log the actual message

Comment (mark-terry) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/EnclaveErrorAcceptanceTest.java:193]: It'd be nice, but I haven't got access to it in this pattern.

File: crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java (hunk 6 above)
Comment (mark-terry) [reply crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java:47]: @pinges  ?

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java (hunk 1 above)
Comment (mark-terry) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/ProcessBesuNodeRunner.java]: Reverted.

File: enclave/src/main/java/org/hyperledger/besu/enclave/EnclaveClientException.java (hunk 8 above)
Comment (mark-terry) [reply enclave/src/main/java/org/hyperledger/besu/enclave/EnclaveClientException.java:17]: Reverted.

**Review by macfarla — APPROVED** [review]
Let's do it

**Review by macfarla — APPROVED** [review]
